src/audio/copier/copier.c

```diff
@@ -208,84 +208,52 @@ static int create_host(struct comp_dev *parent_dev, struct copier_data *cd,
 		       const struct ipc4_copier_module_cfg *copier_cfg,
 		       int dir)
 {
-	struct sof_uuid host = {0x8b9d100c, 0x6d78, 0x418f, {0x90, 0xa3, 0xe0,
-			0xe8, 0x05, 0xd0, 0x85, 0x2b}};
 	struct ipc_config_host ipc_host;
-	const struct comp_driver *drv;
-	struct comp_dev *dev;
 	struct host_data *hd;
 	int ret;
-
-	drv = ipc4_get_drv((uint8_t *)&host);
-	if (!drv)
-		return -EINVAL;
+	enum sof_ipc_frame __sparse_cache in_frame_fmt, out_frame_fmt;
+	enum sof_ipc_frame __sparse_cache in_valid_fmt, out_valid_fmt;
 
 	config->type = SOF_COMP_HOST;
 
-	ret = create_endpoint_buffer(parent_dev, cd, config, copier_cfg, ipc4_gtw_host, false, 0);
-	if (ret < 0)
-		return ret;
+	audio_stream_fmt_conversion(copier_cfg->base.audio_fmt.depth,
+				    copier_cfg->base.audio_fmt.valid_bit_depth,
+				    &in_frame_fmt,
+				    &in_valid_fmt,
+				    copier_cfg->base.audio_fmt.s_type);
+
+	audio_stream_fmt_conversion(copier_cfg->out_fmt.depth,
+				    copier_cfg->out_fmt.valid_bit_depth,
+				    &out_frame_fmt,
+				    &out_valid_fmt,
+				    copier_cfg->out_fmt.s_type);
+
+	if (cd->direction == SOF_IPC_STREAM_PLAYBACK)
+		config->frame_fmt = in_frame_fmt;
+	else
+		config->frame_fmt = out_frame_fmt;
+
+	parent_dev->ipc_config.frame_fmt = config->frame_fmt;
 
 	memset(&ipc_host, 0, sizeof(ipc_host));
 	ipc_host.direction = dir;
 	ipc_host.dma_buffer_size = copier_cfg->gtw_cfg.dma_buffer_size;
 
-	dev = comp_alloc(drv, sizeof(*dev));
-	if (!dev)
-		goto e_buf;
-
-	dev->ipc_config = *config;
-
 	hd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*hd));
 	if (!hd)
-		goto e_dev;
+		return -EINVAL;
 
-	comp_set_drvdata(dev, hd);
-	ret = host_zephyr_new(hd, parent_dev, &ipc_host, dev->ipc_config.id);
+	ret = host_zephyr_new(hd, parent_dev, &ipc_host, config->id);
 	if (ret < 0) {
 		comp_err(parent_dev, "copier: host new failed with exit");
 		rfree(hd);
-		goto e_dev;
-	}
-	dev->state = COMP_STATE_READY;
-
-	list_init(&dev->bsource_list);
-	list_init(&dev->bsink_list);
-
-	if (cd->direction == SOF_IPC_STREAM_PLAYBACK) {
-		comp_buffer_connect(dev, config->core, cd->endpoint_buffer[cd->endpoint_num],
-				    PPL_CONN_DIR_COMP_TO_BUFFER);
-		cd->bsource_buffer = false;
-	} else {
-		comp_buffer_connect(dev, config->core, cd->endpoint_buffer[cd->endpoint_num],
-				    PPL_CONN_DIR_BUFFER_TO_COMP);
-		cd->bsource_buffer = true;
-	}
-
-	cd->converter[IPC4_COPIER_GATEWAY_PIN] =
-		get_converter_func(&copier_cfg->base.audio_fmt,
-				   &copier_cfg->out_fmt,
-				   ipc4_gtw_host, IPC4_DIRECTION(dir));
-	if (!cd->converter[IPC4_COPIER_GATEWAY_PIN]) {
-		comp_err(parent_dev, "failed to get converter for host, dir %d", dir);
-		ret = -EINVAL;
-		goto e_buf;
+		return -EINVAL;
```

Comment:
`return ret`

Reply:
you need to free hd and call host_zephyr_free() here

Reply:
added, @lyakh , at this moment, ret does not make sense.